File: backend/app/services/dashboard_service.py
"""Dashboard aggregation for each role. Cached per user/role."""
from __future__ import annotations

from app.core.redis import Cache
from app.models.enums import ApprovalStatus, JobStatus, UserStatus
from app.repositories.application_repo import ApplicationRepository
from app.repositories.company_repo import CompanyRepository
from app.repositories.job_repo import JobRepository
from app.repositories.misc_repo import InterviewRepository, NotificationRepository
from app.repositories.user_repo import (
    RecruiterProfileRepository,
    StudentProfileRepository,
    UserRepository,
)


class DashboardService:
    def __init__(
        self,
        users: UserRepository,
        students: StudentProfileRepository,
        recruiters: RecruiterProfileRepository,
        companies: CompanyRepository,
        jobs: JobRepository,
        applications: ApplicationRepository,
        interviews: InterviewRepository,
        notifications: NotificationRepository,
    ) -> None:
        self.users = users
        self.students = students
        self.recruiters = recruiters
        self.companies = companies
        self.jobs = jobs
        self.applications = applications
        self.interviews = interviews
        self.notifications = notifications

    async def student(self, user_id: str) -> dict:
        key = f"dash:student:{user_id}"
        if (cached := await Cache.get(key)) is not None:
            return cached

        profile = await self.students.get_by_user(user_id)
        apps = await self.applications.list_for_student(user_id)
        interviews = await self.interviews.list_for_student(user_id)
        active_resume_score = 0
        if profile and profile.active_resume_id:
            from app.core.database import get_database
            from app.repositories.misc_repo import ResumeRepository

            resume = await ResumeRepository(get_database()).get(profile.active_resume_id)
            if resume:
                active_resume_score = resume.analysis.score

        stage_counts: dict[str, int] = {}
        for a in apps:
            stage_counts[a.stage.value] = stage_counts.get(a.stage.value, 0) + 1

        widgets = {
            "profile_completion": profile.profile_strength if profile else 0,
            "placement_readiness": profile.placement_readiness if profile else 0,
            "resume_score": active_resume_score,
            "applications_sent": len(apps),
            "interview_invites": len(interviews),
            "stage_breakdown": stage_counts,
            "skills": profile.skills[:12] if profile else [],
        }
        await Cache.set(key, widgets, ttl=60)
        return widgets

    async def recruiter(self, company_id: str | None) -> dict:
        if not company_id:
            return {"active_jobs": 0, "applicants": 0, "funnel": {}, "interviews": 0}
        key = f"dash:recruiter:{company_id}"
        if (cached := await Cache.get(key)) is not None:
            return cached

        active_jobs = await self.jobs.count(
            {"company_id": company_id, "status": JobStatus.PUBLISHED.value}
        )
        applicants = await self.applications.count({"company_id": company_id})
        funnel = await self.applications.company_stage_counts(company_id)
        interviews = await self.interviews.list_for_company(company_id)

        widgets = {
            "active_jobs": active_jobs,
            "applicants": applicants,
            "funnel": funnel,
            "interview_pipeline": len(interviews),
        }
        await Cache.set(key, widgets, ttl=60)
        return widgets

    async def admin(self) -> dict:
        key = "dash:admin"
        if (cached := await Cache.get(key)) is not None:
            return cached

        total_users = await self.users.count({})
        active_users = await self.users.count({"status": UserStatus.ACTIVE.value})
        students = await self.users.count({"role": "student"})
        recruiters = await self.users.count({"role": "recruiter"})
        jobs_posted = await self.jobs.count({})
        published_jobs = await self.jobs.count({"status": JobStatus.PUBLISHED.value})
        applications = await self.applications.count({})
        pending_companies = await self.companies.count(
            {"approval_status": ApprovalStatus.PENDING.value}
        )

        widgets = {
            "total_users": total_users,
            "active_users": active_users,
            "students": students,
            "recruiters": recruiters,
            "jobs_posted": jobs_posted,
            "published_jobs": published_jobs,
            "applications_submitted": applications,
            "pending_company_approvals": pending_companies,
            # placeholder revenue metric (wire to billing provider in production)
            "mrr_estimate": recruiters * 49,
        }
        await Cache.set(key, widgets, ttl=120)
        return widgets

File: backend/src/services/dashboardService.js
"use strict";

/** Dashboard aggregation for each role. Cached per user/role.
 * Mirrors app/services/dashboard_service.py. */

const {
  User,
  StudentProfile,
  Company,
  Job,
  Application,
  Interview,
  Resume,
} = require("../models");
const { ApprovalStatus, JobStatus, UserStatus, Role } = require("../models/enums");
const { Cache } = require("../config/redis");

async function companyStageCounts(companyId) {
  const rows = await Application.aggregate([
    { $match: { company_id: companyId } },
    { $group: { _id: "$stage", count: { $sum: 1 } } },
  ]);
  const out = {};
  for (const row of rows) out[row._id] = row.count;
  return out;
}

async function student(userId) {
  const key = `dash:student:${userId}`;
  const cached = await Cache.get(key);
  if (cached !== null && cached !== undefined) return cached;

  const profile = await StudentProfile.findOne({ user_id: userId });
  const apps = await Application.find({ student_id: userId }).sort({ created_at: -1 }).limit(50);
  const interviews = await Interview.find({ student_id: userId }).sort({ start_at: 1 }).limit(100);

  let activeResumeScore = 0;
  if (profile && profile.active_resume_id) {
    const resume = await Resume.findById(profile.active_resume_id);
    if (resume) activeResumeScore = resume.analysis.score;
  }

  const stageCounts = {};
  for (const a of apps) stageCounts[a.stage] = (stageCounts[a.stage] || 0) + 1;

  const widgets = {
    profile_completion: profile ? profile.profile_strength : 0,
    placement_readiness: profile ? profile.placement_readiness : 0,
    resume_score: activeResumeScore,
    applications_sent: apps.length,
    interview_invites: interviews.length,
    stage_breakdown: stageCounts,
    skills: profile ? profile.skills.slice(0, 12) : [],
  };
  await Cache.set(key, widgets, 60);
  return widgets;
}

async function recruiter(companyId) {
  if (!companyId) {
    return { active_jobs: 0, applicants: 0, funnel: {}, interviews: 0 };
  }
  const key = `dash:recruiter:${companyId}`;
  const cached = await Cache.get(key);
  if (cached !== null && cached !== undefined) return cached;

  const activeJobs = await Job.countDocuments({
    company_id: companyId,
    status: JobStatus.PUBLISHED,
  });
  const applicants = await Application.countDocuments({ company_id: companyId });
  const funnel = await companyStageCounts(companyId);
  const interviews = await Interview.find({ company_id: companyId })
    .sort({ start_at: 1 })
    .limit(200);

  const widgets = {
    active_jobs: activeJobs,
    applicants,
    funnel,
    interview_pipeline: interviews.length,
  };
  await Cache.set(key, widgets, 60);
  return widgets;
}

async function admin() {
  const key = "dash:admin";
  const cached = await Cache.get(key);
  if (cached !== null && cached !== undefined) return cached;

  const totalUsers = await User.countDocuments({});
  const activeUsers = await User.countDocuments({ status: UserStatus.ACTIVE });
  const students = await User.countDocuments({ role: Role.STUDENT });
  const recruiters = await User.countDocuments({ role: Role.RECRUITER });
  const jobsPosted = await Job.countDocuments({});
  const publishedJobs = await Job.countDocuments({ status: JobStatus.PUBLISHED });
  const applications = await Application.countDocuments({});
  const pendingCompanies = await Company.countDocuments({
    approval_status: ApprovalStatus.PENDING,
  });

  const widgets = {
    total_users: totalUsers,
    active_users: activeUsers,
    students,
    recruiters,
    jobs_posted: jobsPosted,
    published_jobs: publishedJobs,
    applications_submitted: applications,
    pending_company_approvals: pendingCompanies,
    // placeholder revenue metric (wire to billing provider in production)
    mrr_estimate: recruiters * 49,
  };
  await Cache.set(key, widgets, 120);
  return widgets;
}

module.exports = { student, recruiter, admin, companyStageCounts };
